### PR TITLE
Feat constant replication flow

### PR DESCRIPTION
**Pull request by joshuef**
To give an idea, adding:

SendingDataReplicationBatch logs to [here](https://github.com/maidsafe/safe_network/blob/main/sn_node/src/node/api/dispatcher.rs#L563-L567), on main, we see over the course of the network split test, updated w/ same values as churn in this PR

- it's `found: 312 times`. Each one of those is ~50mb. a total of ~15.6gb
- memory jumps high. I have to kill it when it reaches ~700mb per node.

With this PR, over the course of the `churn` test:

- nodes dont go over ~110mb
- it succeeds
- SendingMissingReplicatedData found: 3287 times, 3.2gb... 

This is a slower rate of transfer, which is safer for nodes/systems running nodes, that can more easily be tweaked by the node themselves down the line. (And add in there that there should be less mem per message too!)

### CODE REVIEW

File: sn_node/src/node/core/messaging/handling/update_section.rs [hunk 1]
```diff
@@ -7,43 +7,139 @@
...
+        let adults = self.network_knowledge.adults().await;
+        let adults_names = adults.iter().map(|p2p_node| p2p_node.name()).collect_vec();
+
+        let elders = self.network_knowledge.elders().await;
+        let my_name = self.info.read().await.name();
+
+        // find data targets that are not us.
+        let mut target_member_names = adults_names
+            .into_iter()
+            .sorted_by(|lhs, rhs| my_name.cmp_distance(lhs, rhs))
```
Comment (maqi) [inline sn_node/src/node/core/messaging/handling/update_section.rs:101]: Within an edge case that for a section of prefix(), if our_name is: 0011xxx , other nodes are 0101xxx, 0101xxx, 0101xxx, 0100xxx, 0100xxx .  
Then for a data of 0000xxx,   the closest 3 nodes to the data will be 0011xxx, 0100xxx, 0100xxx .
But the closest 3 nodes to us will be 0101xxx, 0101xxx, 0101xxx,  which missed that data.

**Review by davidrusu — APPROVED** [review]
I like the new churn test, nice work.

We'll def. need to improve on the `SendAnyMissingRelevantData(Vec<ReplicatedDataAddress>)` message as the amount of data grows, perhaps replace the vec with a merkle tree? but anyhow, that's a drop in optimization we can work through later

File: .github/workflows/sn_pr_tests.yml [hunk 2]
```diff
@@ -339,6 +339,26 @@ jobs:
         run: cargo build  --release --bin log_cmds_inspector
         timeout-minutes: 60
 
+      # This starts a 25 node network, we should kill before moving on to split checks
+      - name: Run network churn data integrity test
+        timeout-minutes: 30
+        shell: bash
+        run: cargo run --release --example churn
```
[the remark this answers is not in the document]
Comment (joshuef) [reply .github/workflows/sn_pr_tests.yml:346]: > I thought we prefer to have the CI scripts to be responsible of setup testing environment

Hmm, we haven't had that specified in the past? Also the setup / addition of nodes is kind of crucial to this test?

It cooould go into the section split. Just felt cleaner having those conceptually separate.

File: sn_node/src/node/core/messaging/handling/update_section.rs [hunk 3]
```diff
@@ -7,43 +7,139 @@
...
+
+        trace!(
+            "nearest neighbours for data req: {}: {:?}",
+            target_member_names.len(),
+            target_member_names
+        );
+
+        // also send to our elders in case they are holding but were just promoted
+        for elder in elders {
+            let _existed = target_member_names.insert(elder.name());
```
[the remark this answers is not in the document]
Comment (maqi) [reply sn_node/src/node/core/messaging/handling/update_section.rs:114]: This is to ensure data in newly promoted elders can be replicated.
so, maybe just query the X yongest nodes ?  which they have the highest chance being newly promoted.

**Pull request merged**